### ISSUE TITLE
Transformer T5Encoder get nan in Trainer with 16-mixed precision 

### ISSUE THREAD

**Issue by yanzhenms** [Lightning-AI/lightning#18919]
### Bug description

With plain pytorch and autocast, T5Encoder output values doesn't have nan, but inside a pl trainer 16-mixed context, the output is very unstable and gets nan values very frequently.

### What version are you seeing the problem on?

v2.1

### How to reproduce the bug

```python
import lightning.pytorch as pl
import torch
from transformers import T5EncoderModel, AutoTokenizer

class T5Encoder(pl.LightningModule):
    def __init__(self, encoder):
        super().__init__()
        self.encoder = encoder

    def training_step(self, batch, batch_idx):
        input_ids = batch[0]
        non_pad_mask = input_ids != 0
        hidden = self.encoder(input_ids, attention_mask=non_pad_mask).last_hidden_state
        loss = torch.mean(hidden)
        print(loss)
        return loss
    
    def configure_optimizers(self):
        optimizer = torch.optim.Adam(self.parameters(), lr=1e-4)
        return optimizer
    

def test_wo_pl(model, input_ids):
    with torch.cuda.amp.autocast(True, dtype=torch.float16):
        non_pad_mask = input_ids != 0
        hidden = model(input_ids, attention_mask=non_pad_mask).last_hidden_state
        loss = torch.mean(hidden)
    return loss

def main():
    input_text = [
        "Hello, my dog is cute",
        "Hello, my cat is cute",
        "Oh, my dog is cute",
        "Oh, my dog is cute",
    ]

    tokenizer = AutoTokenizer.from_pretrained("google/byt5-base")
    input_ids = tokenizer(input_text, padding=True, return_tensors="pt").input_ids.cuda()
    model = T5EncoderModel.from_pretrained("google/byt5-base").cuda()

    print("test without lightning")
    for _ in range(10):
        loss = test_wo_pl(model, input_ids)
        print(loss)

    print("test with lightning")
    input_ids = input_ids.repeat(10, 1)
    dataset = torch.utils.data.TensorDataset(input_ids)
    dataloader = torch.utils.data.DataLoader(dataset, batch_size=4, shuffle=False)
    model = T5Encoder(model)
    trainer = pl.Trainer(accelerator='gpu', devices=1, precision='16-mixed', accumulate_grad_batches=10, max_steps=1)
    trainer.fit(model, dataloader)

if __name__ == "__main__":
    main()
```

### Error messages and logs

```
test without lightning
tensor(-0.0009, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(-0.0009, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(-0.0009, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(-0.0009, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(-0.0009, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(-0.0009, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(-0.0009, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(-0.0009, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(-0.0009, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(-0.0009, device='cuda:0', grad_fn=<MeanBackward0>)
test with lightning
Using 16bit Automatic Mixed Precision (AMP)
GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs
You are using a CUDA device ('NVIDIA GeForce RTX 4090') that has Tensor Cores. To properly utilize them, you should set `torch.set_float32_matmul_precision('medium' | 'high')` which will trade-off precision for performance. For more details, read https://pytorch.org/docs/stable/generated/torch.set_float32_matmul_precision.html#torch.set_float32_matmul_precision
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0,1,2,3]

  | Name    | Type           | Params
-------------------------------------------
0 | encoder | T5EncoderModel | 414 M 
-------------------------------------------
414 M     Trainable params
0         Non-trainable params
414 M     Total params
1,658.814 Total estimated model params size (MB)
/data/yanzhen/miniconda3/envs/tortoise/lib/python3.10/site-packages/lightning/pytorch/trainer/connectors/data_connector.py:441: The 'train_dataloader' does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` to `num_workers=127` in the `DataLoader` to improve performance.
/data/yanzhen/miniconda3/envs/tortoise/lib/python3.10/site-packages/lightning/pytorch/loops/fit_loop.py:293: The number of training batches (10) is smaller than the logging interval Trainer(log_every_n_steps=50). Set a lower value for log_every_n_steps if you want to see logs for the training epoch.
Epoch 0:   0%|                                                                                                                                                                        | 0/10 [00:00<?, ?it/s]
tensor(nan, device='cuda:0', grad_fn=<MeanBackward0>)
Epoch 0:  10%|███████████████                                                                                                                                        | 1/10 [00:00<00:01,  6.82it/s, v_num=4]
tensor(nan, device='cuda:0', grad_fn=<MeanBackward0>)
Epoch 0:  20%|██████████████████████████████▏                                                                                                                        | 2/10 [00:00<00:00,  9.28it/s, v_num=4]
tensor(nan, device='cuda:0', grad_fn=<MeanBackward0>)
Epoch 0:  30%|█████████████████████████████████████████████▎                                                                                                         | 3/10 [00:00<00:00, 11.07it/s, v_num=4]
tensor(nan, device='cuda:0', grad_fn=<MeanBackward0>)
Epoch 0:  40%|████████████████████████████████████████████████████████████▍                                                                                          | 4/10 [00:00<00:00, 12.16it/s, v_num=4]
tensor(-0.0007, device='cuda:0', grad_fn=<MeanBackward0>)
Epoch 0:  50%|███████████████████████████████████████████████████████████████████████████▌                                                                           | 5/10 [00:00<00:00, 12.94it/s, v_num=4]
tensor(-0.0009, device='cuda:0', grad_fn=<MeanBackward0>)
Epoch 0:  60%|██████████████████████████████████████████████████████████████████████████████████████████▌                                                            | 6/10 [00:00<00:00, 13.15it/s, v_num=4]
tensor(nan, device='cuda:0', grad_fn=<MeanBackward0>)
Epoch 0:  70%|█████████████████████████████████████████████████████████████████████████████████████████████████████████▋                                             | 7/10 [00:00<00:00, 13.34it/s, v_num=4]
tensor(nan, device='cuda:0', grad_fn=<MeanBackward0>)
Epoch 0:  80%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▊                              | 8/10 [00:00<00:00, 13.41it/s, v_num=4]
tensor(-0.0005, device='cuda:0', grad_fn=<MeanBackward0>)
Epoch 0:  90%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▉               | 9/10 [00:00<00:00, 13.55it/s, v_num=4]
tensor(nan, device='cuda:0', grad_fn=<MeanBackward0>)
Epoch 0: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 10/10 [00:00<00:00, 13.62it/s, v_num=4]`Trainer.fit` stopped: `max_steps=1` reached.
Epoch 0: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 10/10 [00:03<00:00,  2.85it/s, v_num=4]
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version (e.g., 1.5.0):
#- Lightning App Version (e.g., 0.5.2):
#- PyTorch Version (e.g., 2.0):
#- Python version (e.g., 3.9):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
#- Running environment of LightningApp (e.g. local, cloud):
```
</details>

**Comment by yanzhenms** [comment]
seems if from_pretrained returned a model in eval mode, if I call model.train(), it will also return nan without pl Trainer.
